File: src/main/java/no/novari/flyt/history/repository/filters/TimeQueryFilter.java
package no.novari.flyt.history.repository.filters;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.springframework.kafka.support.JavaUtils;

import java.time.OffsetDateTime;
import java.util.Optional;
import java.util.StringJoiner;

@Builder
@EqualsAndHashCode
public class TimeQueryFilter {

    public static final TimeQueryFilter EMPTY =
            new TimeQueryFilter(null, null);

    private final OffsetDateTime latestStatusTimestampMin;
    private final OffsetDateTime latestStatusTimestampMax;

    public Optional<OffsetDateTime> getLatestStatusTimestampMin() {
        return Optional.ofNullable(latestStatusTimestampMin);
    }

    public Optional<OffsetDateTime> getLatestStatusTimestampMax() {
        return Optional.ofNullable(latestStatusTimestampMax);
    }

    @Override
    public String toString() {
        StringJoiner joiner = new StringJoiner(", ", "(", ")");
        JavaUtils.INSTANCE.acceptIfNotNull(
                latestStatusTimestampMin, latestStatusTimestampMin -> joiner.add("latestStatusTimestampMin=" + latestStatusTimestampMin)
        );
        JavaUtils.INSTANCE.acceptIfNotNull(
                latestStatusTimestampMax, latestStatusTimestampMax -> joiner.add("latestStatusTimestampMax=" + latestStatusTimestampMax)
        );
        return joiner.toString();
    }

}

File: src/main/kotlin/no/novari/flyt/history/repository/filters/TimeQueryFilter.kt
package no.novari.flyt.history.repository.filters

import java.time.OffsetDateTime

data class TimeQueryFilter(
    val latestStatusTimestampMin: OffsetDateTime? = null,
    val latestStatusTimestampMax: OffsetDateTime? = null,
) {
    companion object {
        @JvmField
        val EMPTY = TimeQueryFilter()

        @JvmStatic
        fun builder() = Builder()
    }

    class Builder {
        private var latestStatusTimestampMinValue: OffsetDateTime? = null
        private var latestStatusTimestampMaxValue: OffsetDateTime? = null

        fun latestStatusTimestampMin(latestStatusTimestampMin: OffsetDateTime?) =
            apply {
                this.latestStatusTimestampMinValue = latestStatusTimestampMin
            }

        fun latestStatusTimestampMax(latestStatusTimestampMax: OffsetDateTime?) =
            apply {
                this.latestStatusTimestampMaxValue = latestStatusTimestampMax
            }

        fun build() =
            TimeQueryFilter(
                latestStatusTimestampMin = latestStatusTimestampMinValue,
                latestStatusTimestampMax = latestStatusTimestampMaxValue,
            )
    }
}
